test/integration/options/watch.spec.js

```diff
@@ -19,38 +21,156 @@ describe('--watch', function() {
+        });
+    });
+
+    it('reruns test when watched test file is touched', function() {
+      const [mocha, outputPromise] = runMochaJSONWatchAsync(this.testFile, [], {
+        cwd: this.tempDir
+      });
 
-          var sigint = 2;
-          expect(data.code, 'to be', exitStatusBySignal(sigint));
-          done();
-        },
-        spawnOpts
+      return sleep(1000)
+        .then(() => {
+          touchFile(this.testFile);
+          return sleep(1000);
+        })
+        .then(() => {
+          mocha.kill('SIGINT');
+          return outputPromise;
+        })
+        .then(results => {
+          expect(results, 'to have length', 2);
+        });
+    });
+
+    it('reruns test when file matching extension is touched', function() {
+      const watchedFile = path.join(this.tempDir, 'file.xyz');
+      touchFile(watchedFile);
+      const [mocha, outputPromise] = runMochaJSONWatchAsync(
+        this.testFile,
+        ['--extension', 'xyz,js'],
+        {
+          cwd: this.tempDir
+        }
       );
 
-      setTimeout(function() {
-        // Kill the child process
-        mocha.kill('SIGINT');
-      }, 1000);
+      return sleep(1000)
+        .then(() => {
+          touchFile(watchedFile);
+          return sleep(1000);
+        })
+        .then(() => {
+          mocha.kill('SIGINT');
+          return outputPromise;
+        })
+        .then(results => {
+          expect(results, 'to have length', 2);
+        });
+    });
+
+    it('ignores files in "node_modules" and ".git"', function() {
+      const nodeModulesFile = path.join(
+        this.tempDir,
+        'node_modules',
+        'file.xyz'
+      );
+      const gitFile = path.join(this.tempDir, '.git', 'file.xyz');
+
+      touchFile(gitFile);
+      touchFile(nodeModulesFile);
+
+      const [mocha, outputPromise] = runMochaJSONWatchAsync(
+        this.testFile,
+        ['--extension', 'xyz,js'],
+        {
+          cwd: this.tempDir
+        }
+      );
+
+      return sleep(1000)
+        .then(() => {
+          touchFile(gitFile);
+          touchFile(nodeModulesFile);
+        })
+        .then(() => sleep(1000))
+        .then(() => {
+          mocha.kill('SIGINT');
+          return outputPromise;
+        })
+        .then(results => {
+          expect(results, 'to have length', 1);
+        });
     });
   });
 });
+
+/**
+ * Invokes the mocha binary with the `--watch` argument for the given fixture.
+ *
+ * Returns child process and a promise for the test results. The test results
+ * are an array of JSON objects generated by the JSON reporter.
+ *
+ * Checks that the exit code of the mocha command is 130, i.e. mocha was killed
+ * by SIGINT.
+ */
+function runMochaJSONWatchAsync(fixture, args, spawnOpts) {
+  args = ['--watch'].concat(args);
+  const [mocha, mochaDone] = runMochaJSONRawAsync(fixture, args, spawnOpts);
+  const testResults = mochaDone.then(data => {
+    expect(data.code, 'to be', sigintExitCode);
```

Comment:
I think we can skip this.  these tests aren't asserting Mocha exits how we expect; they are `--watch` tests.

Reply:
https://github.com/mochajs/mocha/pull/3915/commits/b9f42a872426cecb1bc72158f49bff09c2c95442